lib/src/generic/interceptors/csrf_interceptor.dart

```diff
@@ -58,9 +60,18 @@ class CsrfInterceptor extends Interceptor {
   Future<Context> onOutgoing(Provider provider, Context context) async {
     // Inject CSRF token into headers.
     if (context is HttpContext) {
-      if (!context.request.headers.containsKey(_header)) {
+      if (context.request.headers.containsKey(_header)) {
+        if (context.request.headers[_header] == null) {
+          throw new ArgumentError('CSRF header value can not be null');
+        }
+        ;
+      } else {
         context.request.headers[_header] = token;
       }
+
+      if (csrfNotRequired.contains(context.method)) {
+        context.request.headers.remove(_header);
+      }
```

Comment:
I think this could be ordered a bit better. Right now for GET/HEAD/OPTIONS requests you're adding the token and then removing it immediately, which is a bit unnecessary. Additionally, it's not harmful for the CSRF token to be on GET/HEAD/OPTIONS requests, it's just not required. So I'd opt to keep the header there if it was set manually, otherwise it could be confusing to a consumer trying to set the header manually and not seeing it show up on the actual request. 

How about putting another check around line 69 where you set the token (if not already set) and only do it if it's not one of `['GET', 'HEAD', 'OPTIONS']`